src/inventory/slot_mapping.rs

```diff
@@ -0,0 +1,196 @@
+use crate::inventory::{Item, Slot};
+use crate::render::hud::Hud;
+use crate::render::inventory::InventoryWindow;
+use crate::render::Renderer;
+use crate::ui::{Container, VAttach};
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use parking_lot::RwLock;
+
+/// A mapping of each slot and it's position on the screen. This makes the job
+/// of rendering and locating slots easier for inventories. This class only
+/// expects positions in pixels relative to the top-left corner of the window.
+pub struct SlotMapping {
+    /// A mapping of slot_id to (position, slot). The position is the distance
+    /// between the window's top-left corner, and the slot's top-left corner in
+    /// pixels. This doesn't include the border frame around each slot, just
+    /// the 16x16 content.
+    slots: HashMap<u16, ((i32, i32), Slot)>,
```

Comment:
is there a reason this is a HashMap instead of a Vec?

Reply:
Okay, fair enough, the question now becomes, do we always just have an offset for slot positions, or do we sometimes have weird numbering without apparent structure that's out of scope for offsetting? If we just have an offset it might be worth exploring if we could just store said offset inside the slot mapping and apply the offset for every get call and such, because hashmaps will be way more expensive than offsetting (if offsetting is a viable option)

Reply:
Okay, that's fair, a HashMap is justified then

Reply:
I've added a comment.